core-bundle/src/Resources/contao/drivers/DC_Table.php

```diff
@@ -3983,7 +3983,7 @@ protected function generateTree($table, $id, $arrPrevNext, $blnHasSorting, $intM
 		}
 
 		$label = preg_replace('/\(\) ?|\[] ?|{} ?|<> ?/', '', $label);
-		$isVisibleRootTrailPage = \in_array($id, $this->visibleRootTrails);
+		$isVisibleRootTrailPage = $needsCheck ? \in_array($id, $this->visibleRootTrails) : \in_array($objRow->pid, $this->visibleRootTrails);
```

Comment:
Without this check, Contao will display buttons for articles on trail pages that end in a 403 error. @Toflar Not sure if that is the correct fix?

Reply:
I don't understand the change at all. Why do you want to change it?

Reply:
<img width="867" alt="" src="https://user-images.githubusercontent.com/1192057/145564659-c5f835f0-f082-4f3e-bdf2-4b79258a5e96.png">

These two are root trail pages. They are not mounted, so when the user clicks the edit button, it will trigger a 403 exception. After the change, these pages will not have buttons anymore.

<img width="866" alt="" src="https://user-images.githubusercontent.com/1192057/145564776-6673c3ea-09cb-418e-85ba-f083c6507bfc.png">

Reply:
One is `tl_page` and one is `tl_article`. The permission is configured on `tl_page`. So you need to make sure you take the parent ID there, I see.
Hmm, I see the problem but I don't think this is the right solution. I mean, especially because the naming is super confusing now. `$needsCheck` should be renamed then, wdyt?

Reply:
I am only reusing the `$needsCheck` variable because it is already there. 😄 We might as well duplicate the if condition or come up with something more sophisticated.

Reply:
Yeah I just want to make sure we still understand the code in the future :)

Reply:
Improved in 0a30a0026a9881aa5d16e2625ad2e846542ecfba.